widgetssdk/src/main/java/com/glia/widgets/chat/adapter/CustomMessageAdapter.java

```diff
@@ -0,0 +1,66 @@
+package com.glia.widgets.chat.adapter;
+
+import static com.glia.widgets.chat.adapter.ChatAdapter.CUSTOM_MESSAGE_TYPE;
+
+import android.view.LayoutInflater;
+import android.view.ViewGroup;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import com.glia.androidsdk.chat.ChatMessage;
+import com.glia.widgets.UiTheme;
+import com.glia.widgets.chat.adapter.holder.CustomMessageViewHolder;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class CustomMessageAdapter {
+    @VisibleForTesting
+    final Map<Integer, Integer> viewTypeMap = new HashMap<>();
+
+    @Nullable
+    abstract public Integer getItemViewType(ChatMessage message);
+
+    @NonNull
+    abstract public CustomMessageViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
+                                                               @NonNull LayoutInflater inflater,
+                                                               @NonNull UiTheme uiTheme,
+                                                               int viewType);
+
+    @Nullable
+    public final Integer getChatAdapterViewType(ChatMessage chatMessage) {
+        Integer customMessageViewType = getItemViewType(chatMessage);
+        if (customMessageViewType != null) {
+            if (!viewTypeMap.containsKey(customMessageViewType)) {
+                viewTypeMap.put(customMessageViewType, CUSTOM_MESSAGE_TYPE + viewTypeMap.size());
+            }
+            return viewTypeMap.get(customMessageViewType);
+        }
+        return null;
+    }
+
+    @Nullable
+    final Integer getCustomMessageViewTypeMap(int chatAdapterViewType) {
```

Comment:
NIT: Maybe It makes sense to change `viewTypeMap` to `SparseIntArray`, `SparseArrays` are optimized for android and this method will look something like this 
```
int index = viewTypeMap.indexOfValue(chatAdapterViewType);
        return viewTypeMap.keyAt(index); 
```

Reply:
Good idea.

Reply:
Yep, with `viewTypeMap.indexOfValue(chatAdapterViewType);` we are getting `index` of `entry`, and `viewTypeMap.keyAt(index);` will return `key at index`, which is key of `chatAdapterViewType`.

Reply:
Yeah, it works. But I have trouble with unit tests for it. It's because of the error "Method SparseArray ... not mocked". The default implementation doesn't have logic. And I can't set up frameworks like Robolectric to work with the module.
Any ideas?

Reply:
As far as I know, the only way is to use `SparseArrayCompat`